core/src/main/java/org/elasticsearch/plugins/SearchPlugin.java

```diff
@@ -0,0 +1,191 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.NamedWriteable;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.io.stream.Writeable.Reader;
+import org.elasticsearch.common.lucene.search.function.ScoreFunction;
+import org.elasticsearch.common.xcontent.XContent;
+import org.elasticsearch.index.query.functionscore.ScoreFunctionBuilder;
+import org.elasticsearch.index.query.functionscore.ScoreFunctionParser;
+import org.elasticsearch.search.aggregations.bucket.significant.heuristics.SignificanceHeuristic;
+import org.elasticsearch.search.aggregations.bucket.significant.heuristics.SignificanceHeuristicParser;
+import org.elasticsearch.search.aggregations.pipeline.movavg.models.MovAvgModel;
+import org.elasticsearch.search.aggregations.pipeline.movavg.models.MovAvgModel.AbstractModelParser;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.highlight.Highlighter;
+import org.elasticsearch.search.suggest.Suggester;
+
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+
+/**
+ * Plugin for extending search time behavior.
+ */
+public interface SearchPlugin {
+    /**
+     * The new {@link ScoreFunction}s defined by this plugin.
+     */
+    default List<ScoreFunctionSpec<?>> getScoreFunctions() {
+        return emptyList();
+    }
+    /**
+     * The new {@link SignificanceHeuristic}s defined by this plugin.
+     */
+    default List<SignificanceHeuristicSpec> getSignificanceHeuristics() {
+        return emptyList();
+    }
+    /**
+     * The new {@link MovAvgModel}s defined by this plugin.
+     */
+    default List<MovAvgModelSpec> getMovingAverageModels() {
+        return emptyList();
+    }
+    /**
+     * The new {@link FetchSubPhase}s defined by this plugin.
+     */
+    default List<FetchSubPhase> getFetchSubPhases(FetchPhaseConstructionContext context) {
+        return emptyList();
+    }
+    /**
+     * Get the {@link Highlighter}s defined by this plugin.
+     */
+    default Map<String, Highlighter> getHighlighters() {
+        return emptyMap();
+    }
+    /**
+     * The new {@link Suggester}s defined by this plugin.
+     */
+    default Map<String, Suggester<?>> getSuggesters() {
+        return emptyMap();
+    }
+
+    public class ScoreFunctionSpec<T extends ScoreFunctionBuilder<T>> extends SearchPluginSpec<T, ScoreFunctionParser<T>> {
```

Comment:
Now that I'm falling asleep I'm not sure why I made these classes. I'll have to think about that again. Maybe not needed.